Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp

```diff
@@ -12639,65 +12639,99 @@ IGNORE_CLANG_WARNINGS_END
-            bucket = m_out.add(buffer, m_out.shl(m_out.zeroExt(index, Int64), m_out.constInt32(getLSBSet(sizeof(WeakMapBucket<WeakMapBucketDataKey>)))));
-        }
+            LValue bucket;
 
-        LValue bucketKey = m_out.load64(bucket, m_heaps.WeakMapBucket_key);
-        m_out.branch(m_out.equal(key, bucketKey), unsure(continuation), unsure(notEqualValue));
+            if (m_node->child1().useKind() == WeakMapObjectUse) {
+                static_assert(hasOneBitSet(sizeof(WeakMapBucket<WeakMapBucketDataKeyValue>)), "Should be a power of 2");
+                bucket = m_out.add(buffer, m_out.shl(m_out.zeroExt(index, Int64), m_out.constInt32(getLSBSet(sizeof(WeakMapBucket<WeakMapBucketDataKeyValue>)))));
+            } else {
+                static_assert(hasOneBitSet(sizeof(WeakMapBucket<WeakMapBucketDataKey>)), "Should be a power of 2");
+                bucket = m_out.add(buffer, m_out.shl(m_out.zeroExt(index, Int64), m_out.constInt32(getLSBSet(sizeof(WeakMapBucket<WeakMapBucketDataKey>)))));
+            }
 
-        m_out.appendTo(notEqualValue, loopAround);
-        m_out.branch(m_out.isNull(bucketKey), unsure(continuation), unsure(loopAround));
+            LValue bucketKey = m_out.load64(bucket, m_heaps.WeakMapBucket_key);
+            m_out.branch(m_out.equal(key, bucketKey), unsure(continuation), unsure(notEqualValue));
 
-        m_out.appendTo(loopAround, continuation);
-        m_out.addIncomingToPhi(unmaskedIndex, m_out.anchor(m_out.add(index, m_out.int32One)));
-        m_out.jump(loopStart);
+            m_out.appendTo(notEqualValue, loopAround);
+            m_out.branch(m_out.isNull(bucketKey), unsure(continuation), unsure(loopAround));
 
-        m_out.appendTo(continuation, lastNext);
-        LValue result;
-        if (m_node->child1().useKind() == WeakMapObjectUse)
-            result = m_out.load64(bucket, m_heaps.WeakMapBucket_value);
-        else
-            result = bucketKey;
-        setJSValue(result);
+            m_out.appendTo(loopAround, continuation);
+            m_out.addIncomingToPhi(unmaskedIndex, m_out.anchor(m_out.add(index, m_out.int32One)));
+            m_out.jump(loopStart);
+
+            m_out.appendTo(continuation, lastNext);
+            LValue result;
+            if (m_node->child1().useKind() == WeakMapObjectUse)
+                result = m_out.load64(bucket, m_heaps.WeakMapBucket_value);
+            else
+                result = bucketKey;
+            setJSValue(result);
+            break;
+        }
+        case CellUse: {
+            LValue key = lowCell(m_node->child2());
+            if (m_node->child1().useKind() == WeakMapObjectUse)
+                setJSValue(vmCall(Int64, operationWeakMapGet, m_vmValue, weakMap, key, hash));
+            else
+                setJSValue(vmCall(Int64, operationWeakSetGet, m_vmValue, weakMap, key, hash));
+            break;
+        }
```

Comment:
Why slow path here? Why can't we run the above path?

Reply:
Yeah, removed.